pkg/dbadmin/mysqladmin/admin.go

```diff
@@ -265,12 +267,124 @@ func (mdba *MySQLDbAdmin) GetNextIDs(tableNames []dbadmin.TableName) (map[dbadmi
+	switch constraintType {
+	case dbadmin.NotNullConstraint:
+		// Find out if the table already has a not null constraint on all of the named columns
+		query, args, err := sqlx.In("SHOW COLUMNS FROM `?` WHERE Field IN (?);", tableName, columns)
+		if err != nil {
+			return true, fmt.Errorf("Unable to prepare query to read table definition: %w", err)
+		}
+		query = mdba.handle.Rebind(query)
+
+		var fieldDefinitions []struct {
+			Field string
+			Type  string
+			Null  string
+		}
+		if err := mdba.handle.Select(&fieldDefinitions, query, args...); err != nil {
+			return true, wrap(err)
+		}
+
+		fieldNullableMap := make(map[string]string, len(fieldDefinitions))
+		for _, fieldDefinition := range fieldDefinitions {
+			fieldNullableMap[fieldDefinition.Field] = fieldDefinition.Null
+		}
+
+		for _, columnToCheck := range columns {
+			definition, ok := fieldNullableMap[columnToCheck]
+			if !ok {
+				// Will definitely fail because the column doesn't exist
+				return true, nil
+			}
+
+			if definition == "YES" {
+				// The table currently allows for nulls in the column, and we're
+				// trying to make it not nullable. Check if there are any nulls
+				// in that column.
+				query := "SELECT * FROM `?` WHERE `?` NOT NULL LIMIT 1;"
+				_, err := mdba.handle.QueryRowx(query, tableName, columnToCheck).SliceScan()
+				if err != sql.ErrNoRows {
+					if err != nil {
+						return true, wrap(err)
+					}
+
+					// A value was returned, meaning there are null rows in this column
+					return true, nil
+				}
+			}
+		}
+		return false, nil
+	case dbadmin.UniqueConstraint:
+		listUniqueIndexQuery := "SHOW INDEX FROM `?` WHERE Non_unique = 0;"
+
+		var uniqueIndexRows []struct {
+			KeyName    string `db:"Key_name"`
+			ColumnName string `db:"Column_name"`
+		}
+		if err := mdba.handle.Select(&uniqueIndexRows, listUniqueIndexQuery, tableName); err != nil {
+			return true, wrap(err)
+		}
+
+		columnsByIndex := make(map[string]mapset.Set)
+		for _, uniqueIndexRow := range uniqueIndexRows {
+			indexColumnSet, ok := columnsByIndex[uniqueIndexRow.KeyName]
+			if !ok {
+				indexColumnSet = mapset.NewSet()
+				columnsByIndex[uniqueIndexRow.KeyName] = indexColumnSet
+			}
+
+			indexColumnSet.Add(uniqueIndexRow.ColumnName)
+		}
+
+		newConstraintColumnSet := mapset.NewSet()
+		for _, columnName := range columns {
+			newConstraintColumnSet.Add(columnName)
+		}
+		for _, existingIndexColumnSet := range columnsByIndex {
+			if existingIndexColumnSet.Equal(newConstraintColumnSet) {
+				// We can add this constraint because there is already a unique
+				// index over the same columns
+				return false, nil
+			}
+		}
+
+		// There is no existing index, we must check if there is some value in
+		// the database which violates the constraint
+		query, args, err := sqlx.In("SELECT COUNT(*) FROM `?` GROUP BY ? HAVING COUNT(*) > 1 LIMIT 1;", tableName, columns)
```

Comment:
I think this is probably reasonable on certain size tables, but I wonder if there's a size table where you really don't want a query like this to run. Think 200M+ rows.

Reply:
Right, I was trying to figure out how to handle the interplay between large table and constraint failure. It should be prevented due to the table's size.